tests/unit/build_tests/test_classgenerator.py

```diff
@@ -82,6 +82,58 @@ def test_no_generators(self):
         self.assertTrue(hasattr(cls, '__init__'))
 
 
+class TestPostInitGetClass(TestCase):
+    def setUp(self):
+        # self.bar_spec = GroupSpec(
+        #     doc='A test group specification with a data type',
+        #     data_type_def='Bar',
+        #     datasets=[
+        #         DatasetSpec(
+        #             doc='a dataset',
+        #             dtype='int',
+        #             name='data',
+        #             attributes=[AttributeSpec(name='attr1', doc='an integer attribute', dtype='int')]
+        #         )
+        #     ])
+        # specs = [self.bar_spec]
+        # containers = {'Bar': Bar}
+        # from hdmf.common import get_type_map
+        # self.type_map = get_type_map()
+        # self.spec_catalog = self.type_map.namespace_catalog.get_namespace(CORE_NAMESPACE).catalog
+        spec = GroupSpec(
+            doc='A test group specification with a data type',
+            data_type_def='Baz',
+            attributes=[
+                AttributeSpec(name='attr1', doc='a int attribute', dtype='int')
+            ]
+        )
+
+        spec_catalog = SpecCatalog()
+        spec_catalog.register_spec(spec, 'test.yaml')
+        namespace = SpecNamespace(
+            doc='a test namespace',
+            name=CORE_NAMESPACE,
+            schema=[{'source': 'test.yaml'}],
+            version='0.1.0',
+            catalog=spec_catalog
+        )
+        namespace_catalog = NamespaceCatalog()
+        namespace_catalog.add_namespace(CORE_NAMESPACE, namespace)
+        self.type_map = TypeMap(namespace_catalog)
+
+
+    def test_post_init(self):
+        def post_init_method(self, **kwargs):
+            attr1 = kwargs['attr1']
+            if attr1<10:
+                msg = "attr1 should be >=10"
+                raise ValueError(msg)
+
+        cls = self.type_map.get_dt_container_cls('Baz', CORE_NAMESPACE, post_init_method)
```

Comment:
I think this approach binds the method to the class rather than the instance.
```python
cls.post_init_method = tp.MethodType(post_init_method, cls) # set as bounded method
```
acts on the class and creates a classmethod, so the first argument is set to the class instead of the instance.

If your method was instead:
```python
    def post_init_method(self, **kwargs):
            self.attr1 = "hi"  # do anything that modifies `self` like change a dtype or set an attribute
            # this will actually set a class variable "attr1" that is shared across all classes and is distinct from the instance variable "attr1"
```

I believe if you call 
```python
obj = cls(name='instance', attr1=9)
print(obj.attr1)  # should still be 9
```

I don't think you need this line:
```python
cls.post_init_method = tp.MethodType(post_init_method, cls) # set as bounded method
```

Reply:
It does bind the function as a method to the class. My thought was that if we run a post init, a person would want to be able to call help(someclass) and see that this method is included in the class. Even though it binds to the class, it still calls it for each instance.

Reply:
> encapsulate the post init method with the class

I think `cls.post_init_method = post_init_method` (monkey patching) might be sufficient and create a method that is not a class method

Reply:
I think we need to establish what we want.

- Add a post init (when provided) such that the kwargs dict of an instance of the class is validated us that post init.

That means we need the class to always have that post init method. This means we need to bind the function as a method to the class. This is that:

```
cls.post_init_method = tp.MethodType(post_init_method, cls) # bounded method
```
We can go one of two ways:
1 . Keep that line above. What this will do is make the function a bounded method. This means when called, it will provide self as a positional arg. This requires that the user defined the post init as 
```
def foo(self, **kwargs):
```
If they instead do `def foo(**kwargs)`, then an error will return. It will say foo does not take positional arguments but 1 was provided. This is because it is being given "self", because we assigned it as a bounded method.

2. We can do `cls.post_init_method = post_init_method`. This assigns the method to the class. It does not take self. It does not take in cls. It is just a method that takes in kwargs. This is also fine. I think this edge out the first idea because users won't have to define "self" in the class definition parameters.  

Thoughts? @rly

Reply:
I actually don't think we need to make it a bounded method. There's no benefit to using self since it is only operating on kwargs. However, if we want to expand the scope so that the post_init does something with the instance variables then that is a different conversation.

Reply:
> However, if we want to expand the scope so that the post_init does something with the instance variables then that is a different conversation.

Generally, I think the expectation would be that a `post_init` method would be able to act on the object since the purpose of `__init__` is to initialize objects of a class. I would assume a main use-case for `post_init` in an extension is to add validation checks to make sure instance variables are set to valid values and set default values for instance variables if they are not set, or even add new instance variables.

Reply:
Makes sense to me. I believe the request was phrased to operate on the kwargs dict, which is why I raised the question of scope.

Reply:
> I believe the request was phrased to operate on the kwargs dict

Acting on the `kwargs` I don't think is useful in a `post_init` setting, since at that point the `kwargs` have already been used in `__init__` to set instance variables.

Reply:
I agree - it's useful to act on `self` in `post_init`. I believe that is doable through monkeypatching with `cls.post_init_method = post_init_method` and that that binds the method to the class without making the class a classmethod, though I am not totally familiar with the use of bound vs unbound(?) methods.

```python
>>> class A:
...     def __init__(self, arg1):
...         self.arg1 = arg1
... 
>>> def modify_arg1(self):
...     self.arg1 = self.arg1 + 1
... 
>>> A.modify_arg1 = modify_arg1
>>> dir(A)
['__class__', '__delattr__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', '__weakref__', 'modify_arg1']
>>> a = A(10)
>>> a.modify_arg1()
>>> a.arg1
11
>>> a.modify_arg1
<bound method modify_arg1 of <__main__.A object at 0x1007b76d0>>
>>> type(a.modify_arg1)
<class 'method'>
>>> import types
>>> type(a.modify_arg1) == types.MethodType
True
```

Reply:
I believe we are all on the same idea. I'll finalize this PR tomorrow.

Reply:
Bound method just means it is tied to an instance and the first argument will be self. Both of our ways do the same thing.